packages/flutter/lib/src/widgets/display_feature_sub_screen.dart

```diff
@@ -0,0 +1,157 @@
+import 'package:flutter/rendering.dart';
+import 'package:vector_math/vector_math_64.dart';
+
+import 'basic.dart';
+import 'framework.dart';
+import 'media_query.dart';
+import 'overlay.dart';
+
+/// Positions [child] such that it avoids overlapping any [DisplayFeature] that
+/// splits the screen into sub-screens.
+///
+/// After determining the sub-screens, the closest one to [anchorPoint] is used
+/// render the [child].
+///
+/// If no [anchorPoint] is provided, then [Directionality] is used. If no
+/// [Directionality] and no [anchorPoint] are provided, then the top-left screen
+/// is used.
+///
+/// See also:
+///
+///  * [showDialog], which is a way to display a DialogRoute.
+///  * [showCupertinoDialog], which displays an iOS-style dialog.
+class DisplayFeatureSubScreen extends StatelessWidget {
+  /// Creates a widget that positions its child so that it avoids display features.
+  const DisplayFeatureSubScreen({
+    Key? key,
+    required this.child,
+    this.anchorPoint,
+  }) : super(key: key);
+
+  /// The child that will avoid the display features.
+  final Widget child;
+
+  /// The anchor point used to pick the closest area that has no display features.
+  /// `Offset(0,0)` is the top-left corner of the available screen space. For
+  /// a dual-screen device, this is the top-left corner of the left screen.
+  final Offset? anchorPoint;
+
+  @override
+  Widget build(BuildContext context) {
+    final List<Rect> safeAreas = _safeAreasInNavigator(context);
+    final Rect safeArea = anchorPoint == null
+        ? _firstSafeArea(safeAreas, context)
+        : _closestToAnchorPoint(safeAreas, _finiteOffset(anchorPoint!));
+
+    return Align(
+      alignment: Alignment.topLeft,
+      child: Padding(
+        padding: EdgeInsets.only(left: safeArea.left, top: safeArea.top),
+        child: SizedBox(
+          width: safeArea.width,
+          height: safeArea.height,
+          child: child,
+        ),
+      ),
+    );
+  }
+
+  static Rect _firstSafeArea(List<Rect> safeAreas, BuildContext context) {
+    final TextDirection? textDirection = Directionality.maybeOf(context);
+    if (textDirection == TextDirection.rtl)
+      return _closestToAnchorPoint(safeAreas, const Offset(double.maxFinite, 0));
+    else
+      return _closestToAnchorPoint(safeAreas, Offset.zero);
+  }
+
+  static Rect _closestToAnchorPoint(List<Rect> safeAreas, Offset anchorPoint) {
+    return safeAreas.fold(safeAreas.first, (Rect previousValue, Rect element) {
+      final double previousDistance = (previousValue.center - anchorPoint).distanceSquared;
+      final double elementDistance = (element.center - anchorPoint).distanceSquared;
+      if (previousDistance < elementDistance)
+        return previousValue;
+      else
+        return element;
+    });
+  }
+
+  static List<Rect> _safeAreasInNavigator(BuildContext context) {
+    final RenderObject? renderObject = Overlay.of(context)?.context.findRenderObject();
```

Comment:
Just by reading the documentation I am surprised that this interacts with Overlay at all?